File: src/cpp/bridge.cc
#include "bridge.h"

namespace webrtc {

std::unique_ptr<webrtc::InterArrivalDelta> new_inter_arrival_delta() {
    return std::make_unique<webrtc::InterArrivalDelta>(webrtc::TimeDelta::Millis(5));
}

bool ComputeDeltas(
    webrtc::InterArrivalDelta& self,
    uint64_t send_time_us,
    uint64_t arrival_time_us,
    uint64_t system_time_us,
    uint64_t packet_size,
    int64_t& send_time_delta_us,
    int64_t& arrival_time_delta_us,
    uint64_t& packet_size_delta
) {
    TimeDelta send_delta = TimeDelta::Zero();
    TimeDelta recv_delta = TimeDelta::Zero();
    int size_delta = 0;

    auto computed_deltas = self.ComputeDeltas(
        webrtc::Timestamp::Micros(send_time_us),
        webrtc::Timestamp::Micros(arrival_time_us),
        webrtc::Timestamp::Micros(system_time_us),
        packet_size,
        &send_delta,
        &recv_delta,
        &size_delta
    );

    send_time_delta_us = send_delta.us();
    arrival_time_delta_us = recv_delta.us();
    packet_size_delta = size_delta;

    return computed_deltas;
};

}  // namespace bridge

File: src/bridge.rs
pub use self::ffi::*;

#[cxx::bridge(namespace = "webrtc")]
pub mod ffi {
    #[derive(Debug, Eq, Hash, PartialEq)]
    #[repr(i32)]
    pub enum BandwidthUsage {
        kBwNormal,
        kBwUnderusing,
        kBwOverusing,
        kLast,
    }

    unsafe extern "C++" {
        include!("bridge.h");

        pub type InterArrivalDelta;

        pub fn new_inter_arrival_delta() -> UniquePtr<InterArrivalDelta>;

        // This function returns true if a delta was computed, or false if the current
        // group is still incomplete or if only one group has been completed.
        //
        // `send_time` is the send time.
        // `arrival_time` is the time at which the packet arrived.
        // `packet_size` is the size of the packet.
        // `timestamp_delta` (output) is the computed send time delta.
        // `arrival_time_delta` (output) is the computed arrival-time delta.
        // `packet_size_delta` (output) is the computed size delta.
        //
        // bool ComputeDeltas(
        //      Timestamp send_time,            // packet_feedback.sent_packet.send_time,
        //      Timestamp arrival_time,         // packet_feedback.receive_time,
        //      Timestamp system_time,          // at_time
        //      size_t packet_size,             // packet_size.bytes(),
        //      TimeDelta* send_time_delta,     // &send_delta,
        //      TimeDelta* arrival_time_delta,  // &recv_delta,
        //      int* packet_size_delta          // &size_delta
        // );
        pub fn ComputeDeltas(
            iad: Pin<&mut InterArrivalDelta>,
            send_time_us: u64,
            arrival_time_us: u64,
            system_time_us: u64,
            packet_size_bytes: u64,
            send_time_delta_us: &mut i64,
            arrival_time_delta_us: &mut i64,
            packet_size_delta: &mut u64,
        ) -> bool;

    }
}

unsafe impl Send for InterArrivalDelta {}
